prime-router/docs/design/proposals/0024-original-message-passthrough.md

```diff
@@ -0,0 +1,168 @@
+      information from the FHIR bundle when preparing to send it if a FHIR receiver is configured to receive it.
+    * *Con*: This approach limits the possible formats passed through; only CSV and HL7v2 are possible.
+
+* **Add feature to the UP translate step: retrieve blobURL from report_file DB table**:
+
+  In this approach, the process of creating the FHIR bundle is not altered in any way. Instead, the `translate` step is
+  altered to determine if original message passthrough is configured. If so, the `translate` step can then walk through
+  the `report_file` database table in a similar manner as done in the Submission History API to retrieve the blobURL and
+  the externalName of the original HL7v2 file and make a copy of it to send to the receiver.
+    * As information is not being added to the FHIR bundle, the determination that original passthrough is desired must
+      be made another way. This can take several forms, such as introducing a new `Report.Format` (e.g. `HL7_ORIGINAL`)
+      or having the `translate` function retrieve sender settings for the original submission.
+
+    * *Pro*: This approach potentially has the least amount of modifications to existing processes and makes use of
+      information that is already stored in the database. There is no need to store additional information about the
+      submission to facilitate the passthrough feature.
+    * *Pro*: There is no risk of leaking private information about ReportStream's storage as the FHIR bundle is not
+      modified.
+    * *Pro*: There is no limitation on incoming file format.
+    * *Con*: This approach is highly reliant on the processes performed at the `receive` pipeline step remaining as they
+      are now. Any refactoring of this code must take original message passthrough into consideration.
+
+## Limitations
+
+* Filtering can still occur as a routing process, but the nature of sending an unmodified HL7v2 file will mean any
+  conditions that would have been eliminated by filtering would be sent regardless.
+* `MERGE` batching cannot be performed for original passthrough HL7v2 files.
+* Filenames provided by the sender must be unique. New submissions with an existing filename should be rejected.
+* Filenames in the Azure blobstore currently reflect pipeline specific information to more easily identify when and why
+  a file was created.
+    * There is currently a limitation that files must have unique names, and blobURLs cannot be reused.
+    * External filenames are in many cases based on the blobURL presently.
+
+## Feature Design
+
+After performing analysis of the possible implementations above, the **Add feature to the UP translate step: retrieve
+blobURL from report_file DB table** implementation was selected.
+
+The overall design of the original passthrough feature is as follows:
+
+* A new `Topic` will be used for original passthrough items. This guarantees that data that must be handled via original
```

Comment:
I'm warming up to the idea of using Topics for the solution here. Instead of creating a whole new topic for this, what if we add a flag to all topics that indicate whether to essentially skip the translate step or not? It would just be another param akin to isUniversalPipeline:

```
/**
 * A submission with topic FULL_ELR will be processed using the full ELR pipeline (fhir engine), submissions
 * from a sender with topic COVID_19 will be processed using the covid-19 pipeline.
 */
enum class Topic(@JsonValue val jsonVal: String, val isUniversalPipeline: Boolean = true) {
    FULL_ELR("full-elr", true),
    ETOR_TI("etor-ti", true),
    ELR_ELIMS("elr-elims", true),
    COVID_19("covid-19", false),
    MONKEYPOX("monkeypox", false),
    TEST("test", false),
}
```

Then you can just use `Topic.sendOriginal` in translate function for your switch?

The reason I'm thinking more this way is because I'm thinking more about the existential question: "What is a topic?" and its been basically turning into a **contract** for a certain use case. For example, each Topic now indicates which pipeline it will use. So, when you select "FULL_ELR", you as the sender and receiver are both acknowledging and agreeing to the rules associated with that topic. So I think continuing to develop topics in this way (by adding another param called send_original) makes sense. The current topic "ELR_ELIMS" would be updated to have this set to true, the other topics will have it as false. If a sender and receiver both select this topic, they are agreeing to send/receive CDC ETOR ELIMS data from the STARLIMS system with the understanding the original message will be sent and transforms are disabled -- that makes sense to me?

Reply:
If we go this route, that changes our "assumptions" a bit:

Original wording:

> This design is intended for a specific sender. Although the design can be expanded to encompass additional senders as needed, at this time it is less disruptive to assume this sender's requirements are unique. It should also be assumed that the long term goal for all senders on ReportStream is to eventually utilize ReportStream's filtering and routing
processes instead of bypassing these features.

Rough proposed wording:

> While this design is intended to support a specific use case, CDC ELIMS ETOR, it should be easily applicable to other use cases as well and fit into the existing architecture of the pipeline.

Thoughts:

I disagreed with a lot of the language in the original wording. If one use case is asking for this functionality, other use cases may ask for additional functionality as well, either the same or different as this. Thus, it becomes important to start thinking about how can we build this out while allowing us to further expand it in the future in a maintainable way. I'm not sure Topics in the way I propose we use them is the final solution here, but I think its a good first step.